3_harmonize/src/clean_wqp_data.R

```diff
@@ -19,54 +17,44 @@
-#' used to identify a duplicate record. Duplicate records are defined as those 
-#' that share the same value for each column within `duplicate_definition`. By 
-#' default, a record will be considered duplicated if it shares the same 
-#' organization, site id, date, time, characteristic name, and sample fraction. 
-#' However, these options can be customized by passing a vector of column names 
-#' to the argument `duplicate_definition`.
-#' @param remove_duplicated_rows logical; should duplicated records be omitted
-#' from the cleaned dataset? Defaults to TRUE. 
+#' @param remove_duplicated_records logical; should records that are exactly-
+#' duplicated across all columns be omitted from the dataset? Defaults to TRUE. 
 #' 
 #' @returns 
 #' Returns a formatted and harmonized data frame containing data downloaded from 
 #' the Water Quality Portal, where each row represents a unique data record.
 #' 
-clean_wqp_data <- function(wqp_data, char_names_crosswalk,
+clean_wqp_data <- function(wqp_data, 
+                           char_names_crosswalk,
                            commenttext_missing = c('analysis lost', 'not analyzed', 
                                                    'not recorded', 'not collected', 
                                                    'no measurement taken'),
-                           duplicate_definition = c('OrganizationIdentifier',
-                                                    'MonitoringLocationIdentifier',
-                                                    'ActivityStartDate', 
-                                                    'ActivityStartTime.Time',
-                                                    'CharacteristicName', 
-                                                    'ResultSampleFractionText'),
-                           remove_duplicated_rows = TRUE){
+                           remove_duplicated_records = TRUE){
 
   # Clean data and assign flags if applicable
   wqp_data_clean <- wqp_data %>%
     # harmonize characteristic names by assigning a common parameter name
     # to the groups of characteristics supplied in `char_names_crosswalk`.
     left_join(y = char_names_crosswalk, by = c("CharacteristicName" = "char_name")) %>%
     # flag true missing results
-    flag_missing_results(., commenttext_missing) %>%
-    # flag duplicate records
-    flag_duplicates(., duplicate_definition) %>%
-    {if(remove_duplicated_rows){
-      remove_duplicates(., duplicate_definition)
-    } else {.}
-    }
+    flag_missing_results(., commenttext_missing)
   
-  # Inform the user what we found for duplicated rows
-  if(remove_duplicated_rows){
-    message(sprintf(paste0("Removed %s duplicated records."), 
-                    nrow(wqp_data) - nrow(wqp_data_clean)))
+  # Check that no rows were omitted when applying QC flags
+  if(nrow(wqp_data_clean) != nrow(wqp_data)){
+    stop("Records were unintentionally removed during the data flagging step!")
```

Comment:
If I received this as an error message, I would not know what to do next. What sort of action are we hoping for here?

Reply:
I included this as an emergency brake of sorts, but this is a great point and could just be confusing! After thinking about this a bit more - I cannot think of any case where records would be omitted by the above code chunk, since this is using a left join (rather than an inner join, for example). 

However, it would be easy enough to accidentally duplicate (add) rows depending on how the user formats their config file. The `clean_wqp_data()` function assumes that each characteristic name in the `char_names_crosswalk` df corresponds to one "parameter" value (but not more than one!). To give an example, if a user formats the config file in `1_inventory/cfg` to have parameters/headers "nitrate" and "nitrogen" that both contain the characteristic name `"Nitrate-N"`, then we may accidentally duplicate all records with that characteristic name in the data flagging step. Maybe this error isn't even necessary since more recent versions of `dplyr` already warn us about multiple matches in joins. But a more useful error message might be:

```suggestion
  # Check that the number of records is unchanged after applying QC flags
  if(nrow(wqp_data_clean) > nrow(wqp_data)){
    stop(paste0("Records were unintentionally duplicated during the data flagging ", 
               "step. In the `char_names_crosswalk` table, check that each ",
               "'char_name' corresponds with one unique value for 'parameter'."))
  }
```

Reply:
I think that is a great! I like error message that are actionable :)

Reply:
Agreed! Thanks for the nudge 😉 I've made this change in https://github.com/USGS-R/ds-pipelines-targets-example-wqp/pull/112/commits/6589f6444dd53c68e7a34ece2c5bc2bb8c33d00f